File: leetcode/medium/45_Jump_Game_II.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int s = 0;
        int e = nums[0];
        int choose = 1;
        int size = nums.size();
        if (size == 1) {
            return 0;
        }
        while (e < size - 1) {
            int max_jump = 0;
            for (int i = s; i <= e; i++) {
                if (i + nums[i] > max_jump) {
                    max_jump = i + nums[i];
                }
                s=i;
            }
            e = max_jump;
            choose++;
        }
        return choose;
    }
};

File: leetcode/medium/45_Jump_Game_II.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [10001] * n
        dp[n-1] = 0
        for i in range(n-2,-1,-1):
            for j in range(0,nums[i]+1):
                if i+j <= n-1:
                    dp[i] = min(dp[i],1+dp[i+j])
        return dp[0]